### PR TITLE
CP-40029: Add PPI command storge to varstored

### PR DESCRIPTION
**Pull request by xennifer**

### CODE REVIEW

File: io_port.c [hunk 1]
```diff
@@ -110,15 +125,52 @@ io_port_deregister(void)
...
+io_port_read(uint64_t addr, uint64_t size)
+{
+    uint32_t port_index;
+
+    assert(io_port.enable && io_port.addr <= addr
+           && addr < (io_port.addr + io_port.size));
+
+    addr -= io_port.addr;
+
+    if (size <= 4) {
```
Comment (rosslagerwall) [inline io_port.c:159]: Would it not be better to check the alignment like you do for the write side?

File: io_port.c [hunk 2]
```diff
@@ -110,15 +125,52 @@ io_port_deregister(void)
...
+        if (port_index < NUMBER_OF_PORTS && io_port.readl[port_index] != NULL)
+            return (uint64_t) io_port.readl[port_index](addr - port_index * sizeof(int32_t), size);
+        else
+        {
+            DBG("Read from offset %" PRIu64 "not supported.\n", addr);
+            return 0;
+        }
+    }
     else
         DBG("Expected size 4. Got %" PRIu64 ".\n", size);
```
Comment (rosslagerwall) [inline io_port.c:171]: `Expected size <= 4`?

File: io_port.c [hunk 3]
```diff
@@ -44,10 +44,22 @@
 #include <handler.h>
 
 #include "io_port.h"
+#include "ppi.h"
```
Comment (rosslagerwall) [inline io_port.c:47]: This should be `#include <ppi.h>` and placed above by handler.h since it is from the include path rather than a relative path.

**Review by rosslagerwall — CHANGES_REQUESTED** [review]

**Comment by stormi** [comment]
Minor comment: "storge" should probably by "storage" in the commit message.

**Comment by lindig** [comment]
There are some inconsistencies around using `else { stmt; }` and `else stmt;`, also around spacing around relation operators. I would run this trough GNU indent. Didn't spot anything serious, though.

File: io_port.c [hunk 4]
```diff
@@ -110,15 +125,52 @@ io_port_deregister(void)
...
-    if (size == 4)
-        io_port.writel(addr, val);
+    alignment = addr % sizeof(int32_t);
+
+    if (size + alignment <= sizeof(int32_t)) {
+        port_index = addr / sizeof(int32_t);
+        if (port_index < NUMBER_OF_PORTS && io_port.writel[port_index] != NULL)
+           io_port.writel[port_index](addr - port_index * sizeof(int32_t), size, val);
+        else
+           DBG("Write to offset %" PRIu64 "not supported.\n", addr);
```
Comment (edwintorok) [inline io_port.c:143]: should this return an error instead of silently succeeding? Same question about the other else below

File: varstored.c [hunk 5]
```diff
@@ -255,14 +256,19 @@ create_pidfile(const char *path)
...
 {
-    if (ioreq->dir == IOREQ_READ) {
-        DBG("IO request not WRITE. Doing nothing.\n");
-    } else if (ioreq->dir == IOREQ_WRITE) {
-        if (!ioreq->data_is_ptr) {
-            io_port_write(ioreq->addr, ioreq->size, (uint32_t)ioreq->data);
-        } else {
-            assert(0);
-        }
+    if (ioreq->data_is_ptr)
```
Comment (edwintorok) [inline varstored.c:259]: this could be written as `assert(!ioreq->data_is_ptr);`

File: io_port.c (hunk 4 above)
Comment (rosslagerwall) [reply io_port.c:143]: The guest interface to IO ports is just an asm instruction like `outl`. It's not making a hypercall or something so there isn't much scope for returning an error.

File: io_port.c [hunk 6]
```diff
@@ -105,30 +77,127 @@ io_port_deregister(void)
...
+    uint64_t alignment;
+
     assert(io_port.enable && io_port.addr <= addr
            && addr < (io_port.addr + io_port.size));
 
     addr -= io_port.addr;
+    alignment = addr % sizeof(uint32_t);
+
+    if (size == 0) {
+       DBG("Zero sized write to %" PRIu64 "\n", addr);
```
Comment (rosslagerwall) [inline io_port.c:98]: This should probably return here to avoid the callbacks needing to handle weird 0 size writes.

File: io_port.c [hunk 7]
```diff
@@ -105,30 +77,127 @@ io_port_deregister(void)
...
+    uint64_t alignment;
+
+    assert(io_port.enable && io_port.addr <= addr
+           && addr < (io_port.addr + io_port.size));
+
+    addr -= io_port.addr;
+    alignment = addr % sizeof(uint32_t);
+
+    if (size == 0) {
+       DBG("Zero sized read from %" PRIu64 "\n", addr);
```
Comment (rosslagerwall) [inline io_port.c:125]: This should probably return here.

File: io_port.c [hunk 8]
```diff
@@ -105,30 +77,127 @@ io_port_deregister(void)
...
+    if ((address - io_port.addr) & 0x3) {
+        ERR("Failed to register io port handler, address not aligned: %016"PRIx32"\n", address);
+        return -1;
+    }
+    *port = (address - io_port.addr) >> 2;
+    return 0;
+}
+
+int
+register_io_port_readl_handler(uint64_t address, readl_callback_t callback) {
```
Comment (rosslagerwall) [inline io_port.c:161]: `io_port.addr` is `uint32_t` and `address` is uint32_t so I guess it should be here as well?

If you change that then update the logging to use `%08"PRIx32"`.

(And same for `register_io_port_writel_handler` below.)

File: io_port.c [hunk 9]
```diff
@@ -105,30 +77,127 @@ io_port_deregister(void)
...
+
+    return 0;
+}
+
+
+static int
+get_port(uint32_t address, uint32_t *port) {
+
+    if (address < io_port.addr  || address > io_port.addr + io_port.size - sizeof(uint32_t)) {
+        ERR("Failed to register io port handler, address out of range: %016"PRIx32"\n", address);
```
Comment (rosslagerwall) [inline io_port.c:149]: Should be `%08`.

File: io_port.c [hunk 10]
```diff
@@ -105,30 +77,127 @@ io_port_deregister(void)
...
+        return -1;
+    }
+    if ((address - io_port.addr) & 0x3) {
+        ERR("Failed to register io port handler, address not aligned: %016"PRIx32"\n", address);
```
Comment (rosslagerwall) [inline io_port.c:153]: Should be `%08`.

File: io_port.c [hunk 11]
```diff
@@ -105,30 +77,127 @@ io_port_deregister(void)
...
+    }
+    *port = (address - io_port.addr) >> 2;
+    return 0;
+}
+
+int
+register_io_port_readl_handler(uint64_t address, readl_callback_t callback) {
+    uint32_t port;
+    int r;
+    r = get_port(address, &port);
```
Comment (rosslagerwall) [inline io_port.c:164]: As `get_port()` simply either returns success or failure, it's probably better for it to return a bool.

Then you can simplify this to `if (!get_port(address, &port)) ...`

The same goes for the return types of `register_io_port_readl_handler` and `register_io_port_writel_handler`.

File: ppi.c [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+                                       &ppiBufferGuid,
+                                       buf,
+                                       sizeof(buf),
+                                       ATTR_BRNV);
+        if (status != EFI_SUCCESS) {
+            ERR("internal_set_variable returned 0x%016lx\n", status);
+            return false;
+        }
+    } else {
+        free(data);
```
Comment (rosslagerwall) [inline ppi.c:77]: data should only be set if `internal_get_variable()` returns EFI_SUCCESS so surely this should go after the check for EFI_SUCCESS below?

File: io_port.c (hunk 6 above)
Comment (xennifer) [reply io_port.c:98]: Oh, yeah, i ment to put that there.

**Review by rosslagerwall — APPROVED** [review]

File: io_port.c (hunk 8 above)
Comment (xennifer) [reply io_port.c:163]: Made io_port.adder etc; uint64_t

**Pull request merged**